src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -102,7 +102,41 @@ static void uct_cuda_ipc_cache_invalidate_regions(uct_cuda_ipc_cache_t *cache,
               cache->name, from, to);
 }
 
-UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_ipc_cache_map_memhandle,
+ucs_status_t uct_cuda_ipc_unmap_memhandle(void *rem_cache, uintptr_t d_bptr,
+                                          void *mapped_addr, int cache_enabled)
+{
+    uct_cuda_ipc_cache_t *cache = (uct_cuda_ipc_cache_t *) rem_cache;
+    ucs_status_t status         = UCS_OK;
+    ucs_pgt_region_t *pgt_region;
+    uct_cuda_ipc_cache_region_t *region;
+
+    pthread_rwlock_rdlock(&cache->lock);
+    pgt_region = UCS_PROFILE_CALL(ucs_pgtable_lookup, &cache->pgtable, d_bptr);
+    ucs_assert(pgt_region != NULL);
+    region = ucs_derived_of(pgt_region, uct_cuda_ipc_cache_region_t);
+
+    ucs_assert(region->refcount >= 1);
+    region->refcount--;
+
+    /*
+     * check refcount to see if an in-flight transfer is using the same mapping
+     */
+    if (!region->refcount && !cache_enabled) {
+        status = ucs_pgtable_remove(&cache->pgtable, &region->super);
```

Comment:
this operation changes the page table, so we need to use write-lock

Reply:
I'll change to use `pthread_rwlock_wrlock`  in that case.

Reply:
@yosefe I've switched to using wrlock now for this path. I've used `atomic_sub` to update refcount under wrlock. Is this overkill?